In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df=pd.read_csv("/content/IMDb Movies India.csv", encoding='latin1')

In [ ]:
df.head()

Name    Year Duration            Genre  \
0                                         NaN      NaN            Drama   
1  #Gadhvi (He thought he was Gandhi)  (2019)  109 min            Drama   
2                         #Homecoming  (2021)   90 min   Drama, Musical   
3                             #Yaaram  (2019)  110 min  Comedy, Romance   
4                   ...And Once Again  (2010)  105 min            Drama   

   Rating Votes            Director       Actor 1             Actor 2  \
0     NaN   NaN       J.S. Randhawa      Manmauji              Birbal   
1     7.0     8       Gaurav Bakshi  Rasika Dugal      Vivek Ghamande   
2     NaN   NaN  Soumyajit Majumdar  Sayani Gupta   Plabita Borthakur   
3     4.4    35          Ovais Khan       Prateik          Ishita Raj   
4     NaN   NaN        Amol Palekar  Rajat Kapoor  Rituparna Sengupta   

           Actor 3  
0  Rajendra Bhatia  
1    Arvind Jangid  
2       Roy Angana  
3  Siddhant Kapoor  
4      Antara Mali

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15509 entries, 0 to 15508
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      15509 non-null  object 
 1   Year      14981 non-null  object 
 2   Duration  7240 non-null   object 
 3   Genre     13632 non-null  object 
 4   Rating    7919 non-null   float64
 5   Votes     7920 non-null   object 
 6   Director  14984 non-null  object 
 7   Actor 1   13892 non-null  object 
 8   Actor 2   13125 non-null  object 
 9   Actor 3   12365 non-null  object 
dtypes: float64(1), object(9)
memory usage: 1.2+ MB


In [6]:
df.describe()

Rating
count  7919.000000
mean      5.841621
std       1.381777
min       1.100000
25%       4.900000
50%       6.000000
75%       6.800000
max      10.000000

In [7]:
df.isnull().sum()


Name           0
Year         528
Duration    8269
Genre       1877
Rating      7590
Votes       7589
Director     525
Actor 1     1617
Actor 2     2384
Actor 3     3144
dtype: int64

In [8]:
df = df.dropna(subset=['Rating'])


In [9]:
df.isnull().sum()


Name           0
Year           0
Duration    2068
Genre        102
Rating         0
Votes          0
Director       5
Actor 1      125
Actor 2      200
Actor 3      292
dtype: int64

In [10]:
df['Year'] = pd.to_numeric(df['Year'], errors='coerce')



In [14]:

df['Duration'] = df['Duration'].astype(str)
df['Duration'] = df['Duration'].str.replace(' min', '', regex=False)
df['Duration'] = pd.to_numeric(df['Duration'], errors='coerce')
df['Duration'] = df['Duration'].fillna(df['Duration'].median())



In [15]:
df['Year'].dtype


dtype('float64')

In [16]:
df['Year'] = df['Year'].fillna(df['Year'].median())


In [ ]:
df['Duration'] = pd.to_numeric(df['Duration'], errors='coerce')

In [17]:
df['Duration'] = df['Duration'].fillna(df['Duration'].median())


In [19]:
df['Genre'] = df['Genre'].fillna('Unknown')
df['Director'] = df['Director'].fillna('Unknown')
df['Actor 1'] = df['Actor 1'].fillna('Unknown')
df['Actor 2'] = df['Actor 2'].fillna('Unknown')
df['Actor 3'] = df['Actor 3'].fillna('Unknown')


In [21]:
df['Votes'] = df['Votes'].fillna(0)



In [22]:
df.isnull().sum()

Name           0
Year        7919
Duration       0
Genre          0
Rating         0
Votes          0
Director       0
Actor 1        0
Actor 2        0
Actor 3        0
dtype: int64

In [27]:
df.drop(columns=['Year'], inplace=True)


In [28]:
df.isnull().sum()

Name        0
Genre       0
Rating      0
Votes       0
Director    0
Actor 1     0
Actor 2     0
Actor 3     0
dtype: int64

In [29]:
df.drop(columns=['Name'], inplace=True)


In [30]:
X = df.drop('Rating', axis=1)
y = df['Rating']


In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7919 entries, 1 to 15508
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Genre     7919 non-null   object 
 1   Rating    7919 non-null   float64
 2   Votes     7919 non-null   object 
 3   Director  7919 non-null   object 
 4   Actor 1   7919 non-null   object 
 5   Actor 2   7919 non-null   object 
 6   Actor 3   7919 non-null   object 
dtypes: float64(1), object(6)
memory usage: 494.9+ KB


In [32]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for col in X.columns:
    X[col] = le.fit_transform(X[col])


In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


In [34]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(X_train, y_train)


RandomForestRegressor(random_state=42)

In [35]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

y_pred = model.predict(X_test)

print("MAE:", mean_absolute_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("R2 Score:", r2_score(y_test, y_pred))


MAE: 1.0194672664141415
RMSE: 1.280065213279852
R2 Score: 0.11864495437785427


In [36]:
model = RandomForestRegressor(
    n_estimators=300,
    max_depth=15,
    random_state=42
)


In [37]:
X = pd.get_dummies(X, columns=['Genre'])


In [38]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)


LinearRegression()

The model achieved an MAE of approximately 1.02.
While the R² score is relatively low, this is expected due to the categorical nature of the features and the absence of additional influential variables such as budget and reviews.

I built a regression-based Movie Rating Prediction model, performed data preprocessing, feature encoding, trained a Random Forest Regressor, and evaluated it using MAE, RMSE, and R² score

The model converts categorical movie features into numerical form, learns patterns using a Random Forest Regressor, and predicts movie ratings by averaging predictions from multiple decision trees